cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/templates/etl/realtime/sources/JmsSource.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.templates.etl.realtime.sources;
+
+import co.cask.cdap.templates.etl.api.ValueEmitter;
+import co.cask.cdap.templates.etl.api.realtime.RealtimeConfigurer;
+import co.cask.cdap.templates.etl.api.realtime.RealtimeSource;
+import co.cask.cdap.templates.etl.api.realtime.SourceContext;
+import co.cask.cdap.templates.etl.api.realtime.SourceState;
+import co.cask.cdap.templates.etl.realtime.jms.JmsProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.LinkedBlockingQueue;
+import javax.annotation.Nullable;
+import javax.jms.BytesMessage;
+import javax.jms.Connection;
+import javax.jms.ConnectionFactory;
+import javax.jms.Destination;
+import javax.jms.JMSException;
+import javax.jms.Message;
+import javax.jms.MessageConsumer;
+import javax.jms.MessageListener;
+import javax.jms.Session;
+import javax.jms.TextMessage;
+
+/**
+ * <p>
+ * Implementation of CDAP {@link RealtimeSource} that listen to external JMS producer by managing internal
+ * JMS Consumer and send the message as String to the CDAP ETL Template flow via {@link ValueEmitter}
+ * </p>
+ */
+public class JmsSource extends RealtimeSource<String> implements MessageListener {
+  private static final Logger LOG = LoggerFactory.getLogger(JmsSource.class);
+
+  private int jmsAcknowledgeMode = Session.AUTO_ACKNOWLEDGE;
+  private JmsProvider jmsProvider;
+  private final LinkedBlockingQueue<Message> messageQueue = new LinkedBlockingQueue<Message>();
```

Comment:
It could be. I used LinkedBlockingQueue type before because was thinking about calling take() method on that class instead of poll, which can block when no item is in the queue.

Can change it to BlockingQueue if preferable.